include/envoy/upstream/cluster_manager.h

```diff
@@ -309,6 +328,32 @@ class ClusterManager {
   virtual const ClusterRequestResponseSizeStatNames&
   clusterRequestResponseSizeStatNames() const PURE;
   virtual const ClusterTimeoutBudgetStatNames& clusterTimeoutBudgetStatNames() const PURE;
+
+  /**
+   * Request an on-demand discovery of a cluster with a passed name. Passed ODCDS may be used to
+   * perform the discovery process in the main thread if there is no discovery going on for this
+   * cluster. The passed callback will be invoked when the cluster is added and warmed up. It is
+   * expected that the callback will be destroyed when it is invoked. To cancel the discovery,
+   * destroy the returned handle and the callback.
+   *
+   * This function is thread-safe.
+   *
+   * @param weak_odcds is a weak pointer to the ODCDS that will be used for discovery.
+   * @param name is the name of the cluster to be discovered.
+   * @param callback will be called when the discovery is finished.
+   * @return the discovery process handle
+   */
+  virtual ClusterDiscoveryCallbackHandlePtr
+  requestOnDemandClusterDiscovery(OdCdsApiWeakPtr weak_odcds, const std::string& name,
+                                  ClusterDiscoveryCallbackWeakPtr callback) PURE;
+
+  /**
+   * Notifies the threads that were waiting for discovery of a cluster with a passed name. Usually
+   * used to unblock the requests in case of discovery failure.
+   *
+   * @param name is the cluster's name.
+   */
+  virtual void notifyOnDemandCluster(const std::string& name, bool cluster_exists) PURE;
```

Comment:
It seems odd to me that this is a public method? Why is this needed? Shouldn't this be an implementation detail of the internal process that does on demand discovery?

Reply:
I didn't have a clear idea how to signal a failed discovery from ODCDS (not only the case where `onConfigUpdateFailed` is called, but also when requesting a cluster and it's repeatedly absent in response). ODCDS takes `ClusterManager&` as a parameter, not `ClusterManagerImpl`, so I needed to make `notifyOnDemandCluster` a part of `ClusterManager` API. This notification is used to effectively unpause the paused request in the on-demand extension.

I suppose that instead of this `notifyOnDemandCluster` I could install a timer for each discovery process that invokes the callbacks in each thread. That would simplify the ODCDS code and make the `notifyOnDemandCluster` function unnecessary. I only wonder what the timeout should be? Something specified in the on-demand configuration? Or a hardcoded value?

Reply:
I made a change in a commit here: https://github.com/kinvolk/envoy/commit/b94a489c9cbcfedde9a36dd73479ff392f763adb, but I haven't yet pushed it - building it locally and making sure that it works.

Reply:
OK I can take a look at your updated change when you push it. In general I think a callback on failure is fine, but the way I would implement it is to pass a callback object that is also notified on failure. At that point it's an internal implementation detail and you shouldn't need this method.